compiler/rustc_data_structures/src/sync/lock.rs

```diff
@@ -150,24 +162,57 @@ impl LockRaw {
 
     #[inline(always)]
     fn lock(&self) {
+        // SAFETY: This is safe since `self.sync` is used in accordance with the preconditions of
+        // `lock_assume_no_sync` and `lock_assume_sync`.
+        unsafe {
+            if likely(!self.sync) {
+                self.lock_assume_no_sync()
+            } else {
+                self.lock_assume_sync();
+            }
+        }
+    }
+
+    /// This acquires the lock assuming no syncronization is required.
+    ///
+    /// Safety
+    /// This method must only be called if `might_be_dyn_thread_safe` was false on lock creation.
+    #[inline(always)]
+    unsafe fn lock_assume_no_sync(&self) {
         if super::ERROR_CHECKING {
             // We're in the debugging mode, so assert that the lock is not held so we
             // get a panic instead of waiting for the lock.
             assert_eq!(self.try_lock(), true, "lock must not be hold");
         } else {
-            // SAFETY: This is safe since the union fields are used in accordance with `self.sync`.
+            // SAFETY: This is safe since `self.opt.cell` is the union field used due to the
+            // precondition on this function.
             unsafe {
-                if likely(!self.sync) {
-                    if unlikely(self.opt.cell.replace(true)) {
-                        cold_path(|| panic!("lock was already held"))
-                    }
-                } else {
-                    self.opt.lock.lock();
+                if unlikely(self.opt.cell.replace(true)) {
+                    cold_path(|| panic!("lock was already held"))
                 }
             }
         }
     }
 
+    /// This acquires the lock assuming syncronization is required.
+    ///
+    /// Safety
+    /// This method must only be called if `might_be_dyn_thread_safe` was true on lock creation.
+    #[inline(always)]
+    unsafe fn lock_assume_sync(&self) {
```

Comment:
Do we have to add several unsafe functions? We can just do this under `Lock`'s method

Reply:
This keeps the code non-generic and it also makes `LockRaw` more fully featured.

Reply:
I'm not sure if it's worth adding extra unsafe functions, after all, both are to reduce maintenance costs. 
cc @compiler-errors